File: project-{topic}-advanced/src/services/UserService.cpp
```cpp
#include "UserService.hpp"
#include "../utils/PasswordUtils.hpp" // For PWDUtils::hashPassword

UserService::UserService(std::shared_ptr<DatabaseManager> db_manager)
    : db_manager_(db_manager) {}

std::optional<User> UserService::getUserById(int id) {
    Logger::log(LogLevel::DEBUG, "Attempting to retrieve user with ID: " + std::to_string(id));
    auto user = db_manager_->getUserById(id);
    if (!user.has_value()) {
        throw NotFoundException("User with ID " + std::to_string(id) + " not found.");
    }
    return user;
}

std::vector<User> UserService::getAllUsers() {
    Logger::log(LogLevel::DEBUG, "Attempting to retrieve all users.");
    return db_manager_->getAllUsers();
}

User UserService::updateUser(int id, const UserUpdateDTO& user_dto) {
    Logger::log(LogLevel::INFO, "Attempting to update user with ID: " + std::to_string(id));

    std::optional<User> existing_user_opt = db_manager_->getUserById(id);
    if (!existing_user_opt.has_value()) {
        throw NotFoundException("User with ID " + std::to_string(id) + " not found.");
    }

    User updated_user = existing_user_opt.value();

    if (user_dto.username.has_value()) {
        if (db_manager_->getUserByUsername(user_dto.username.value()).has_value() &&
            db_manager_->getUserByUsername(user_dto.username.value())->id.value() != id) {
            throw ConflictException("Username '" + user_dto.username.value() + "' already taken.");
        }
        updated_user.username = user_dto.username.value();
    }
    if (user_dto.email.has_value()) {
        if (db_manager_->getUserByEmail(user_dto.email.value()).has_value() &&
            db_manager_->getUserByEmail(user_dto.email.value())->id.value() != id) {
            throw ConflictException("Email '" + user_dto.email.value() + "' already taken.");
        }
        updated_user.email = user_dto.email.value();
    }
    if (user_dto.password.has_value()) {
        updated_user.password_hash = PWDUtils::hashPassword(user_dto.password.value());
    }
    if (user_dto.role.has_value()) {
        std::string role_str = user_dto.role.value();
        std::transform(role_str.begin(), role_str.end(), role_str.begin(), ::toupper);
        if (role_str == "ADMIN") {
            updated_user.role = UserRole::ADMIN;
        } else if (role_str == "USER") {
            updated_user.role = UserRole::USER;
        } else {
            throw BadRequestException("Invalid role specified: " + user_dto.role.value());
        }
    }

    try {
        db_manager_->updateUser(updated_user);
        Logger::log(LogLevel::INFO, "User with ID " + std::to_string(id) + " updated successfully.");
        return updated_user;
    } catch (const DatabaseException& e) {
        Logger::log(LogLevel::ERROR, "Database error updating user " + std::to_string(id) + ": " + std::string(e.what()));
        throw ServiceException("Failed to update user due to database error.");
    }
}

bool UserService::deleteUser(int id) {
    Logger::log(LogLevel::INFO, "Attempting to delete user with ID: " + std::to_string(id));

    // Check if user exists before attempting to delete
    if (!db_manager_->getUserById(id).has_value()) {
        throw NotFoundException("User with ID " + std::to_string(id) + " not found.");
    }

    try {
        db_manager_->deleteUser(id);
        Logger::log(LogLevel::INFO, "User with ID " + std::to_string(id) + " deleted successfully.");
        return true;
    } catch (const DatabaseException& e) {
        Logger::log(LogLevel::ERROR, "Database error deleting user " + std::to_string(id) + ": " + std::string(e.what()));
        throw ServiceException("Failed to delete user due to database error.");
    }
}

User UserService::createUser(const UserRegisterDTO& register_dto, UserRole role) {
    Logger::log(LogLevel::INFO, "Attempting to create user: " + register_dto.username + " with role: " + userRoleToString(role));

    // Check if username or email already exists
    if (db_manager_->getUserByUsername(register_dto.username).has_value()) {
        throw ConflictException("Username '" + register_dto.username + "' already exists.");
    }
    if (db_manager_->getUserByEmail(register_dto.email).has_value()) {
        throw ConflictException("Email '" + register_dto.email + "' already exists.");
    }

    std::string hashed_password = PWDUtils::hashPassword(register_dto.password);
    User new_user(register_dto.username, register_dto.email, hashed_password, role);

    try {
        int new_user_id = db_manager_->createUser(new_user);
        new_user.id = new_user_id;
        Logger::log(LogLevel::INFO, "User " + new_user.username + " created successfully with ID " + std::to_string(new_user_id));
        return new_user;
    } catch (const DatabaseException& e) {
        Logger::log(LogLevel::ERROR, "Database error during user creation: " + std::string(e.what()));
        throw ServiceException("Failed to create user due to database error.");
    }
}
```

File: project-{topic}-plus/performance_monitor/services/user_service.py
```python
from performance_monitor.extensions import db
from performance_monitor.models import User
from flask import current_app

class UserService:
    """
    Business logic for User management.
    """

    @staticmethod
    def create_user(username, email, password, is_admin=False):
        """Creates a new user."""
        if User.query.filter_by(username=username).first():
            return None, "Username already exists."
        if User.query.filter_by(email=email).first():
            return None, "Email already registered."

        user = User(username=username, email=email, is_admin=is_admin)
        user.set_password(password)
        db.session.add(user)
        try:
            db.session.commit()
            current_app.logger.info(f"User '{username}' created successfully.")
            return user, None
        except Exception as e:
            db.session.rollback()
            current_app.logger.error(f"Error creating user '{username}': {e}")
            return None, "Database error during user creation."

    @staticmethod
    def get_user_by_id(user_id):
        """Retrieves a user by ID."""
        return User.query.get(user_id)

    @staticmethod
    def get_user_by_username(username):
        """Retrieves a user by username."""
        return User.query.filter_by(username=username).first()

    @staticmethod
    def get_all_users():
        """Retrieves all users."""
        return User.query.all()

    @staticmethod
    def update_user(user_id, data):
        """Updates an existing user's information."""
        user = User.query.get(user_id)
        if not user:
            return None, "User not found."

        if 'username' in data and data['username'] != user.username:
            if User.query.filter_by(username=data['username']).first():
                return None, "Username already exists."
            user.username = data['username']
        
        if 'email' in data and data['email'] != user.email:
            if User.query.filter_by(email=data['email']).first():
                return None, "Email already registered."
            user.email = data['email']

        if 'password' in data and data['password']:
            user.set_password(data['password'])
        
        if 'is_admin' in data:
            user.is_admin = data['is_admin']

        try:
            db.session.commit()
            current_app.logger.info(f"User '{user.username}' updated successfully.")
            return user, None
        except Exception as e:
            db.session.rollback()
            current_app.logger.error(f"Error updating user '{user.username}': {e}")
            return None, "Database error during user update."

    @staticmethod
    def delete_user(user_id):
        """Deletes a user by ID."""
        user = User.query.get(user_id)
        if not user:
            return False, "User not found."
        
        db.session.delete(user)
        try:
            db.session.commit()
            current_app.logger.info(f"User '{user.username}' (ID: {user_id}) deleted successfully.")
            return True, None
        except Exception as e:
            db.session.rollback()
            current_app.logger.error(f"Error deleting user '{user.username}' (ID: {user_id}): {e}")
            return False, "Database error during user deletion."

```